In [4]:
from sklearn import cross_validation, datasets, metrics, neighbors 
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from pandas import DataFrame

In [29]:
digits = datasets.load_digits()

digits_frame = DataFrame(digits.data)
digits_frame['target'] = digits.target

print digits_frame.head()

     0    1    2     3     4     5    6    7    8    9   ...     55   56   57  \
0  0.0  0.0  5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   
1  0.0  0.0  0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   
2  0.0  0.0  0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   
3  0.0  0.0  7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0   ...    0.0  0.0  0.0   
4  0.0  0.0  0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   

    58    59    60    61   62   63  target  
0  6.0  13.0  10.0   0.0  0.0  0.0       0  
1  0.0  11.0  16.0  10.0  0.0  0.0       1  
2  0.0   3.0  11.0  16.0  9.0  0.0       2  
3  7.0  13.0  13.0   9.0  0.0  0.0       3  
4  0.0   2.0  16.0   4.0  0.0  0.0       4  

[5 rows x 65 columns]


In [18]:
cancer = datasets.load_breast_cancer()

cancer_frame = DataFrame(cancer.data)
cancer_frame.columns = cancer.feature_names
cancer_frame['target'] = cancer.target
cancer_frame.target = cancer_frame.target.apply(lambda x : cancer.target_names[x])

print cancer_frame.head()

   mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension    ...      worst texture  worst perimeter  \
0                 0.

In [31]:
from sklearn import model_selection as ms
from sklearn import naive_bayes as nb

In [42]:
clf = nb.BernoulliNB()
bernoulli_accuracy_digits = ms.cross_val_score(clf, digits.data, digits.target).mean()
bernoulli_accuracy_cancer = ms.cross_val_score(clf, cancer.data, cancer.target).mean()
print bernoulli_accuracy_digits, bernoulli_accuracy_cancer

0.825823650778 0.627420402859


In [43]:
clf = nb.MultinomialNB()
multinomial_accuracy_digits = ms.cross_val_score(clf, digits.data, digits.target).mean()
multinomial_accuracy_cancer = ms.cross_val_score(clf, cancer.data, cancer.target).mean()
print multinomial_accuracy_digits, multinomial_accuracy_cancer

0.870877148974 0.894579040193


In [44]:
clf = nb.GaussianNB()
gaussian_accuracy_digits = ms.cross_val_score(clf, digits.data, digits.target).mean()
gaussian_accuracy_cancer = ms.cross_val_score(clf, cancer.data, cancer.target).mean()
print gaussian_accuracy_digits, gaussian_accuracy_cancer

0.818600380355 0.936749280609


Логично, что распределение бернулли дало самые плохие результаты, так как оно слишком простое и, видимо, плохо приближает признаки, так как они устроены сложнее.
Логично, что мультиминомиальное распределение дало хороший результат для чисел, так как определить число - это выбрать один вариант из 10, где они между собой особо независимы и можно предположить, что и признаки распределены соответствующим образом (как минимум они целые числа).
Логично, что нормальное распределение лучше всего подошло в случае рака, так как раковая опухоль - явление природное и логично, что ее размер и прочие характеристики распределены нормально.

1. На датасете breast_cancer максимальное качество классфикации вышло - 93.7%
2. На датасете digits максимальное качество классфикации вышло - 87.1%
3. d верно, остальные, вроде, нет